Add ESLint and TypeScript

pmbanugo opened the pull request:
Closes https://github.com/vercel/micro/issues/457

This PR adds TypeScript and rewrites the library code using TypeScript. 

**Highlight:**

- Rewrote the library using TS
- Rewrote the test using TS and node-tap
- Updated the GitHub Actions
- Generate types for the library
- Removes deprecated endpoints
- Drops Node.js 12 support

pmbanugo committed:
chore: add .npmrc

pmbanugo committed:
feat: configure TypeScript and ESLint

pmbanugo committed:
feat!: rewrite modules/files using TypeScript

pmbanugo committed:
chore: export types for the modules

leerob commented on packages/micro/src/bin/micro.ts:
```diff
@@ -0,0 +1,270 @@
+  process.exit(1);
+}
+
+// TODO: consider removing the deprecated endpoints support with a major release?
```
Yeah, this would be a good time to do it - this will be published as a major version.

leerob replied:
We can also drop Node.js 12 support.

pmbanugo replied:
I've removed those and set 14.5 as the minimum Node version

leerob commented on packages/micro/src/bin/micro.ts:
```diff
@@ -0,0 +1,270 @@
+    });
+
+    // `micro` is designed to run only in production, so
+    // this message is perfectly for prod
```
"perfectly for prod" seems like a typo.

leerob commented on packages/micro/src/lib/handler.ts:
```diff
@@ -0,0 +1,30 @@
+  let mod: unknown;
+
+  try {
+    mod = file.endsWith('.mjs') ? await import(file) : await require(file); // Await to support exporting Promises
```
Heads up: I ended up reverting this from the other PR just to make sure we had time to test it out here before the major release.

pmbanugo replied:
I've updated to match the latest main commit. I think it can be done in a different PR.

pmbanugo committed:
feat: change minimum node version to 14.5.0

BREAKING CHANGE: Switched to using ES2020 syntax and minimum Node engine to 14.5.0.

pmbanugo committed:
feat!: removes the deprecated cli options

BREAKING CHANGE: Removes the deprecated options/arguments for starting the server. They are:

- --port (-p)
- --host (-h)
- --unix-socket (-s)

pmbanugo committed:
test!: rewrite test using TypeScript & node-tap

pmbanugo committed:
build: update ci script

pmbanugo commented on test/development.js:
```diff
@@ -1,44 +0,0 @@
-// Packages
```
I didn't add a test for this because the library is only to be used in production. Therefore, I don't find any benefit in checking for `NODE_ENV === 'development'`

The same applies for **production.js**.

the function `sendError` is updated accordingly

pmbanugo commented on packages/micro/src/lib/index.ts:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  if (obj instanceof Stream || readable(obj)) {
+    //TODO: Wouldn't (obj instanceof Stream) be the only check here? Do we specifically need a Readable stream or a Stream object that's not of NodeJS Stream?
```
I'm not sure about this. The comment is for someone here to let me know why we do this. Or discuss later in the future if this is necessary.

pmbanugo committed:
chore: update ESLint config

pmbanugo committed:
ci: run the test script in the right directory

pmbanugo committed:
ci: add test script to root package.json

leerob reviewed (approved):
Will try to get one more review on here tomorrow, but LGTM.

pmbanugo committed:
doc: update usage programmatically
